compiler/src/dotty/tools/dotc/core/Annotations.scala

```diff
@@ -53,33 +56,74 @@ object Annotations {
+          //     that are not on CapSet to `^`. Skolems and capturing types on types
+          //     other than CapSet are not allowed in a retains annotation anyway,
+          //     so the underlying type does not matter. This simplification prevents
+          //     exponential blowup in some cases. See i24556.scala and i24556a.scala.
+          //  3. Drop the annotation entirely if CC is not enabled somehwere.
+
+          def sanitize(tp: Type): Type = tp match
+            case SkolemType(_) =>
+              SkolemType(defn.AnyType)
+            case tp @ AnnotatedType(parent, ann)
+            if ann.symbol.isRetainsLike && parent.typeSymbol != defn.Caps_CapSet =>
+              tp.derivedAnnotatedType(parent, Annotation(defn.RetainsCapAnnot, ann.tree.span))
+            case tp @ OrType(tp1, tp2) =>
+              tp.derivedOrType(sanitize(tp1), sanitize(tp2))
+            case _ =>
+              tp
+
+          def rebuild(tree: Tree, mappedType: Type): Tree = tree match
+            case Apply(fn, Nil) => cpy.Apply(tree)(rebuild(fn, mappedType), Nil)
+            case TypeApply(fn, arg :: Nil) => cpy.TypeApply(tree)(fn, TypeTree(mappedType) :: Nil)
+            case Block(Nil, expr) => rebuild(expr, mappedType)
+
+          if !Feature.ccEnabledSomewhere then
+            EmptyAnnotation // strip retains-like annotations unless capture checking is enabled
+          else
+            val mappedType = sanitize(tm(arg.tpe))
+            if mappedType `eql` arg.tpe then this
+            else derivedAnnotation(rebuild(tree, mappedType))
+
+        case args =>
+          // Checks if `tm` would result in any change by applying it to types
+          // inside the annotations' arguments and checking if the resulting types
+          // are different.
+          val findDiff = new TreeAccumulator[Type]:
+            def apply(x: Type, tree: Tree)(using Context): Type =
+              if tm.isRange(x) then x
+              else
+                val tp1 = tm(tree.tpe)
+                foldOver(if !tp1.exists || tp1.eql(tree.tpe) then x else tp1, tree)
+          val diff = findDiff(NoType, args)
+          if tm.isRange(diff) then EmptyAnnotation
+          else if diff.exists then derivedAnnotation(tm.mapOver(tree))
+          else this
+    end mapWith
 
     /** Does this annotation refer to a parameter of `tl`? */
     def refersToParamOf(tl: TermLambda)(using Context): Boolean =
-      val args = tpd.allArguments(tree)
-      if args.isEmpty then false
-      else tree.existsSubTree:
-        case id: (Ident | This) => id.tpe.stripped match
-          case TermParamRef(tl1, _) => tl eq tl1
-          case _ => false
+      def isLambdaParam(t: Type) = t match
+        case TermParamRef(tl1, _) => tl eq tl1
         case _ => false
+      tpd.allArguments(tree).exists: arg =>
+        if arg.isType then
+          arg.tpe.existsPart(isLambdaParam, stopAt = StopAt.Static)
+        else
+          arg.existsSubTree:
```

Comment:
Great, that probably partially fixes https://github.com/scala/scala3/issues/22008!

Reply:
Ah, good. I had not seen that issue before! I discovered this after a lengthy debug session where I wondered why my changes to `mapWith` would cause lots of errors with orphan parameters in pickling. ~So I assume we can close #22008~? I see that not: Term trees generated by macros might still cause problems.

Reply:
> I discovered this after a lengthy debug session

Sorry for the loss of time; we should probably have added a TODO when we discussed the bug first.

> I see that not: Term trees generated by macros might still cause problems.

Yes, that's a concern.

And outside of macros, are we sure that TermParamRef really can't appear in types of trees other than Ident, This or TypeTree? I don't know how to validate that assumption. Morally, I still feel the safe and correct implementation should be to visit each type part of each tree as in https://github.com/scala/scala3/pull/22001#discussion_r1853079956. But that might be a performance hit.